xdsl/parser.py

```diff
@@ -1,1335 +1,1924 @@
-                 skip_white_space: bool = True) -> str | None:
-        """Get the next n characters (including the current one)"""
-        assert n >= 0
-        if skip_white_space:
-            self.skip_white_space()
-        if self._pos is None:
-            return None
-        if self._pos.idx + n >= len(self.str):
-            return None
-        return self.str[self._pos.idx:self._pos.idx + n]
+        self.last_token = self.next_token(peek=True)
+
+    def save(self) -> save_t:
+        """
+        Create a checkpoint in the parsing process, useful for backtracking
+        """
+        return self.pos, self.break_on
 
-    _T = TypeVar("_T")
+    def resume_from(self, save: save_t):
+        """
+        Resume from a previously saved position.
 
-    def try_parse(self,
-                  parse_fn: Callable[[], _T | None],
-                  skip_white_space: bool = True) -> _T | None:
+        Restores the state of the tokenizer to the exact previous position
         """
-        Wrap a parsing function. If the parsing fails, then return without
-        any change to the current position.
+        self.pos, self.break_on = save
+
+    @contextlib.contextmanager
+    def backtracking(self, region_name: str | None = None):
         """
-        if skip_white_space:
-            self.skip_white_space()
-        start_pos = self._pos
+        This context manager can be used to mark backtracking regions.
+
+        When an error is thrown during backtracking, it is recorded and stored together
+        with some meta information in the history attribute.
+
+        The backtracker accepts the following exceptions:
+         - ParseError: signifies that the region could not be parsed because of (unexpected) syntax errors
+         - BacktrackingAbort: signifies that backtracking was aborted, not necessarily indicating a syntax error
+         - AssertionError: this error should probably be phased out in favour of the two above
+         - EOFError: signals that EOF was reached unexpectedly
+
+        Any other error will be printed to stderr, but backtracking will continue as normal.
+        """
+        save = self.save()
+        starting_position = self.pos
         try:
-            return parse_fn()
-        except ParserError:
-            pass
-        self._pos = start_pos
-        return None
+            yield
+            # clear error history when something doesn't fail
+            # this is because we are only interested in the last "cascade" of failures.
+            # if a backtracking() completes without failre, something has been parsed (we assume)
+            if self.pos > starting_position and self.history is not None:
+                self.history = None
+        except Exception as ex:
+            how_far_we_got = self.pos
+
+            # AssertionErrors act upon the consumed token, this means we only go to the start of the token
+            if isinstance(ex, BacktrackingAbort):
```

Comment:
Why is this different from going back to the start position?

Reply:
I removed this, it was outdated (and the comments were misleading)